In [1]:
import pandas as pd
import numpy as np

In [5]:
salary_train=pd.read_csv('SalaryData_Train (2).csv')
salary_test=pd.read_csv('SalaryData_Test (2).csv')

In [7]:
salary_train.head()

age          workclass   education  educationno        maritalstatus  \
0   39          State-gov   Bachelors           13        Never-married   
1   50   Self-emp-not-inc   Bachelors           13   Married-civ-spouse   
2   38            Private     HS-grad            9             Divorced   
3   53            Private        11th            7   Married-civ-spouse   
4   28            Private   Bachelors           13   Married-civ-spouse   

           occupation    relationship    race      sex  capitalgain  \
0        Adm-clerical   Not-in-family   White     Male         2174   
1     Exec-managerial         Husband   White     Male            0   
2   Handlers-cleaners   Not-in-family   White     Male            0   
3   Handlers-cleaners         Husband   Black     Male            0   
4      Prof-specialty            Wife   Black   Female            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            13   United-States   <=50K  
2            0            40   United-States   <=50K  
3            0            40   United-States   <=50K  
4            0            40            Cuba   <=50K

In [10]:
salary_test.head()

age   workclass      education  educationno        maritalstatus  \
0   25     Private           11th            7        Never-married   
1   38     Private        HS-grad            9   Married-civ-spouse   
2   28   Local-gov     Assoc-acdm           12   Married-civ-spouse   
3   44     Private   Some-college           10   Married-civ-spouse   
4   34     Private           10th            6        Never-married   

           occupation    relationship    race    sex  capitalgain  \
0   Machine-op-inspct       Own-child   Black   Male            0   
1     Farming-fishing         Husband   White   Male            0   
2     Protective-serv         Husband   White   Male            0   
3   Machine-op-inspct         Husband   Black   Male         7688   
4       Other-service   Not-in-family   White   Male            0   

   capitalloss  hoursperweek          native  Salary  
0            0            40   United-States   <=50K  
1            0            50   United-States   <=50K  
2            0            40   United-States    >50K  
3            0            40   United-States    >50K  
4            0            30   United-States   <=50K

In [11]:
string_columns = ["workclass","education","maritalstatus","occupation","relationship","race","sex","native"]

In [12]:
##Preprocessing the data. As, there are categorical variables
from sklearn.preprocessing import LabelEncoder
number = LabelEncoder()
for i in string_columns:
    salary_train[i]= number.fit_transform(salary_train[i])
    salary_test[i]=number.fit_transform(salary_test[i])

In [14]:
salary_train

age  workclass  education  educationno  maritalstatus  occupation  \
0       39          5          9           13              4           0   
1       50          4          9           13              2           3   
2       38          2         11            9              0           5   
3       53          2          1            7              2           5   
4       28          2          9           13              2           9   
...    ...        ...        ...          ...            ...         ...   
30156   27          2          7           12              2          12   
30157   40          2         11            9              2           6   
30158   58          2         11            9              6           0   
30159   22          2         11            9              4           0   
30160   52          3         11            9              2           3   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  \
0                 1     4    1         2174            0            40   
1                 0     4    1            0            0            13   
2                 1     4    1            0            0            40   
3                 0     2    1            0            0            40   
4                 5     2    0            0            0            40   
...             ...   ...  ...          ...          ...           ...   
30156             5     4    0            0            0            38   
30157             0     4    1            0            0            40   
30158             4     4    0            0            0            40   
30159             3     4    1            0            0            20   
30160             5     4    0        15024            0            40   

       native  Salary  
0          37   <=50K  
1          37   <=50K  
2          37   <=50K  
3          37   <=50K  
4           4   <=50K  
...       ...     ...  
30156      37   <=50K  
30157      37    >50K  
30158      37   <=50K  
30159      37   <=50K  
30160      37    >50K  

[30161 rows x 14 columns]

In [15]:
salary_test

age  workclass  education  educationno  maritalstatus  occupation  \
0       25          2          1            7              4           6   
1       38          2         11            9              2           4   
2       28          1          7           12              2          10   
3       44          2         15           10              2           6   
4       34          2          0            6              4           7   
...    ...        ...        ...          ...            ...         ...   
15055   33          2          9           13              4           9   
15056   39          2          9           13              0           9   
15057   38          2          9           13              2           9   
15058   44          2          9           13              0           0   
15059   35          3          9           13              2           3   

       relationship  race  sex  capitalgain  capitalloss  hoursperweek  \
0                 3     2    1            0            0            40   
1                 0     4    1            0            0            50   
2                 0     4    1            0            0            40   
3                 0     2    1         7688            0            40   
4                 1     4    1            0            0            30   
...             ...   ...  ...          ...          ...           ...   
15055             3     4    1            0            0            40   
15056             1     4    0            0            0            36   
15057             0     4    1            0            0            50   
15058             3     1    1         5455            0            40   
15059             0     4    1            0            0            60   

       native  Salary  
0          37   <=50K  
1          37   <=50K  
2          37    >50K  
3          37    >50K  
4          37   <=50K  
...       ...     ...  
15055      37   <=50K  
15056      37   <=50K  
15057      37   <=50K  
15058      37   <=50K  
15059      37    >50K  

[15060 rows x 14 columns]

In [16]:
colnames = salary_train.columns
colnames

Index(['age', 'workclass', 'education', 'educationno', 'maritalstatus',
       'occupation', 'relationship', 'race', 'sex', 'capitalgain',
       'capitalloss', 'hoursperweek', 'native', 'Salary'],
      dtype='object')

In [17]:
len(colnames)

14

There are a total of 13 columns including 'Salary'

In [18]:
x_train = salary_train[colnames[0:13]]
y_train = salary_train[colnames[13]]
x_test = salary_test[colnames[0:13]]
y_test = salary_test[colnames[13]]

In [19]:
## building naive bayes models

from sklearn.naive_bayes import MultinomialNB as MB
from sklearn.naive_bayes import GaussianNB as GB

In [22]:
## Building Multinomial Naives Bayes model for training data

classifier_mb = MB()
classifier_mb.fit(x_train,y_train)
pred_mb = classifier_mb.predict(x_train)
accuracy_mb_train = np.mean(pred_mb == y_train)
accuracy_mb_train


0.7729186698053778

The Multinomial Naives Bayes model running the train data shows accuracy of 77.3%

In [21]:
pd.crosstab(pred_mb, y_train)

Salary   <=50K   >50K
row_0                
 <=50K   21717   5913
 >50K      936   1595

In [25]:
## Building Multinomial Naives Bayes model for test data

pred_mb_test = classifier_mb.predict(x_test)
accuracy_mb_test = np.mean(pred_mb_test == y_test)
accuracy_mb_test

0.7749667994687915

The Multinomial Naives Bayes model running the test data shows accuracy of 77.5%

In [26]:
pd.crosstab(pred_mb_test, y_test)

Salary   <=50K   >50K
row_0                
 <=50K   10891   2920
 >50K      469    780

In [27]:
## Building Gaussian Naives Bayes model for training data

classifier_gb = GB()
classifier_gb.fit(x_train, y_train)
pred_gb = classifier_gb.predict(x_train)
accuracy_gb_train = np.mean(pred_gb == y_train)
accuracy_gb_train

0.7953317197705646

The Gaussian Naives Bayes model running the train data shows accuracy of 79.5%

In [28]:
pd.crosstab(pred_gb,y_train)

Salary   <=50K   >50K
row_0                
 <=50K   21505   5025
 >50K     1148   2483

In [29]:
## Building Gaussian Naives Bayes model for test data

pred_gb_train = classifier_gb.predict(x_test)
accuracy_gb_test = np.mean(pred_gb_train == y_test)
accuracy_gb_test

0.7946879150066402

The Gaussian Naives Bayes model running the test data shows accuracy of 79.46%

In [30]:
pd.crosstab(pred_gb_train,y_test)

Salary   <=50K   >50K
row_0                
 <=50K   10759   2491
 >50K      601   1209

# We've achieve higher accuracy when using Gaussian Naves Bayes model (79.5% accuracy for training data). Therefore, it is best to use GB instead of MB for this prediction